### PR TITLE
Add closeable hook to VertxInternal to allow extension to make cleanu…

### PR DESCRIPTION
**Pull request by vietj**
…p when Vertx closes

We do have this feature on Context that is useful for cleaning up Deployments, but we don't have a global one for Vertx. It is currently needed in vertx-shell that deploys a _fake_ verticle to monitor when Vert.x is closed which is somewhat an hack and hard to test (https://github.com/vert-x3/vertx-shell/blob/master/src/main/java/io/vertx/ext/shell/command/impl/CommandRegistryImpl.java#L59)

This allows vertx extension to be aware when Vert.x closes.

### CODE REVIEW

**Comment by vietj** [comment]
please review this feature

**Comment by vietj** [comment]
@cescoffier @pmlopes @alexlehm or anyone else please review

File: src/main/java/io/vertx/core/impl/CloseHooks.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  private final Logger log;
+  private volatile boolean closeHooksRun;
+  private Set<Closeable> closeHooks;
+
+  CloseHooks(Logger log) {
+    this.log = log;
+  }
+
+  void add(Closeable hook) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: synchronization issue here. You may instantiate the set several times.

File: src/main/java/io/vertx/core/impl/CloseHooks.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    if (closeHooks == null) {
+      // Has to be concurrent as can be removed from non context thread
+      closeHooks = new ConcurrentHashSet<>();
+    }
+    closeHooks.add(hook);
+  }
+
+  void remove(Closeable hook) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: synchronization issue here.

File: src/main/java/io/vertx/core/impl/CloseHooks.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+  }
+
+  void run(Handler<AsyncResult<Void>> completionHandler) {
+    if (closeHooksRun) {
+      // Sanity check
+      throw new IllegalStateException("Close hooks already run");
+    }
+    closeHooksRun = true;
+    if (closeHooks != null && !closeHooks.isEmpty()) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: synchronization issue if an `add` happens at the same time.

File: src/main/java/io/vertx/core/impl/CloseHooks.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  synchronized void remove(Closeable hook) {
+    if (closeHooks != null) {
+      closeHooks.remove(hook);
+    }
+  }
+
+  void run(Handler<AsyncResult<Void>> completionHandler) {
+    Set<Closeable> copy = null;
+    synchronized (this) {
+      if (closeHooksRun) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: if synchronized, the volatile flag is not required anymore.

File: src/main/java/io/vertx/core/impl/VertxInternal.java [hunk 5]
```diff
@@ -120,4 +122,9 @@
    */
   AddressResolverGroup<InetSocketAddress> nettyAddressResolverGroup();
 
+  @GenIgnore
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/VertxInternal.java]: javadoc ?

File: src/main/java/io/vertx/core/impl/VertxInternal.java [hunk 6]
```diff
@@ -120,4 +122,9 @@
...
+  void addCloseHook(Closeable hook);
+
+  @GenIgnore
+  void removeCloseHook(Closeable hook);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/VertxInternal.java]: javadoc ?

File: src/main/java/io/vertx/core/impl/CloseHooks.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        // Sanity check
+        throw new IllegalStateException("Close hooks already run");
+      }
+      closeHooksRun = true;
+      if (closeHooks != null && !closeHooks.isEmpty()) {
+        // Must copy before looping as can be removed during loop otherwise
+        copy = new HashSet<>(closeHooks);
+      }
+    }
+    if (copy != null && !closeHooks.isEmpty()) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: closeHooks is not guarded

Comment (vietj) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: right

File: src/main/java/io/vertx/core/impl/CloseHooks.java (hunk 4 above)
Comment (vietj) [inline src/main/java/io/vertx/core/impl/CloseHooks.java]: +1

**Comment by cescoffier** [comment]
👍

**Pull request merged**